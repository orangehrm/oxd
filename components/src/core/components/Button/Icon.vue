<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <button
    :disabled="disabled"
    v-if="withContainer"
    type="button"
    :class="classes"
    @click="onClick"
  >
    <oxd-icon :class="{'--disabled': disabled}" :name="name" />
  </button>
  <oxd-icon
    v-else
    :name="name"
    :class="{'oxd-icon-button__icon': true, '--disabled': disabled}"
    @click="onClick"
  />
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Icon from '@ohrm/oxd/core/components/Icon/Icon.vue';

export default defineComponent({
  name: 'oxd-icon-button',

  components: {
    'oxd-icon': Icon,
  },

  props: {
    name: {
      type: String,
      required: true,
    },
    withContainer: {
      type: Boolean,
      default: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },

  emits: ['click'],

  computed: {
    classes(): object {
      return {
        'oxd-icon-button': true,
      };
    },
  },

  methods: {
    onClick(e: Event) {
      e.preventDefault();
      this.$emit('click', e);
    },
  },
});
</script>

<style src="./icon.scss" lang="scss" scoped></style>
