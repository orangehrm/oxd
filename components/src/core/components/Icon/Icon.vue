<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<script lang="ts">
import {computed, defineComponent, h, unref} from 'vue';
import {TYPES, TYPE_BOOTSTRAP} from './types';
import * as SVGs from './map';
import SVGLoader from '@ohrm/oxd/core/components/Icon/SVGLoader.vue';

export default defineComponent({
  name: 'oxd-icon',

  props: {
    name: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      default: TYPE_BOOTSTRAP,
      validator: (value: string) => {
        return TYPES.indexOf(value) !== -1;
      },
    },
  },

  setup(props) {
    const classes = computed(() => {
      return {
        'oxd-icon': true,
        [`bi-${props.name}`]: props.type === TYPE_BOOTSTRAP,
      };
    });

    return () =>
      props.type === TYPE_BOOTSTRAP
        ? h('i', {class: unref(classes)})
        : h(
            SVGLoader,
            {class: unref(classes)},
            {default: () => h(SVGs[props.name])},
          );
  },
});
</script>

<style src="./icon.scss" lang="scss" scoped></style>
