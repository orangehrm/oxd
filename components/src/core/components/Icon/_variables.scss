$oxd-icon-font-color: $oxd-interface-gray-darken-1-color !default;
$oxd-icon-font-color--disabled: $oxd-interface-gray-lighten-2-color !default;

$oxd-icon-font-size--extra-large: 18px !default;
$oxd-icon-font-size--large: 16px !default;
$oxd-icon-font-size--medium: 14px !default;
$oxd-icon-font-size--small: 12px !default;
$oxd-icon-font-size--extra-small: 10px !default;

$oxd-svg-icon-font-size--extra-large: 22px !default;
$oxd-svg-icon-font-size--large: 20px !default;
$oxd-svg-icon-font-size--medium: 18px !default;
$oxd-svg-icon-font-size--small: 16px !default;
$oxd-svg-icon-font-size--extra-small: 14px !default;
$oxd-svg-icon-font-size--xx-small: 12px !default;
$oxd-svg-icon-font-size--xxx-small: 10px !default;

$oxd-help-popup-icon-color: $oxd-interface-gray-color !default;
$oxd-help-popup-icon-hover-color: $oxd-primary-one-color !default;
$oxd-help-popup-icon-z-index: $oxd-tooltip-font-z-index !default;

$oxd-help-popup-icon-text-font-family: $oxd-font-family !default;
$oxd-help-popup-icon-text-font-size: 0.75rem !default;
$oxd-help-popup-icon-text-font-style: normal !default;
$oxd-help-popup-icon-text-font-weight: 600 !default;
$oxd-help-popup-icon-text-color: $oxd-interface-gray-darken-1-color !default;
$oxd-help-popup-icon-text-background: $oxd-white-color !default;

$oxd-help-popup-icon-text-max-width: 200px !default;
$oxd-help-popup-icon-text-padding: 0.8rem !default;
$oxd-help-popup-icon-text-border-radius: 1rem !default;
$oxd-help-popup-icon-text-box-shadow: 0 0 35px 0 rgba(0, 0, 0, 0.18) !default;
