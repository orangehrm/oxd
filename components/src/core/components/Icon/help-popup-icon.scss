@import '../../../styles';

.oxd-svg-icon {
  color: $oxd-interface-gray-color;
  &:hover {
    color: $oxd-primary-one-color;
  }
}

[help-tooltip] {
  position: relative;
  cursor: pointer;
}

/* Make help icon show over the tooltip */
[help-tooltip]:hover > svg {
  z-index: $oxd-tooltip-font-z-index;
}

/* Applies to all tooltips */
[help-tooltip]::after {
  text-transform: none;
  font-size: 0.625rem;
  font-style: normal;
  font-weight: 400;
  line-height: 1;

  pointer-events: none;
  position: absolute;
  display: none;
  opacity: 0;

  content: attr(help-tooltip);

  font-family: $oxd-font-family;
  text-align: left;

  max-width: 200px;
  width: max-content;
  overflow: auto;
  white-space: pre-wrap;
  padding: 0.8rem;
  border-radius: 1rem;
  box-shadow: 0 0 35px 0 rgba(0, 0, 0, 0.18);
  background: #fff;
  color: $oxd-interface-gray-darken-1-color;
  z-index: $oxd-tooltip-font-z-index;
  top: 50%;
  left: 50%;
}

/* Make the tooltips respond to hover */
[help-tooltip]:hover::after,
[help-tooltip]:focus-visible::after {
  display: block;
  animation: tooltips-horz 300ms ease-out forwards;
}

/* don't show empty tooltips */
[help-tooltip='']::after {
  display: none !important;
}

/* KEYFRAMES */
@keyframes tooltips-vert {
  to {
    opacity: .9;
  }
}

@keyframes tooltips-horz {
  to {
    opacity: .9;
  }
}
