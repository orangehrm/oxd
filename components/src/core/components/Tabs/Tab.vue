<template>
  <div v-show="selectedTabId === tab.id">
    <slot></slot>
  </div>
</template>
<script lang="ts">
import {defineComponent, inject, PropType} from 'vue';
import {Tab} from './types';

export default defineComponent({
  name: 'oxd-tab',
  props: {
    tab: {
      type: Object as PropType<Tab>,
    },
  },
  setup: function() {
    const selectedTabId = inject('selectedTabId');
    return {
      selectedTabId,
    };
  },
});
</script>
