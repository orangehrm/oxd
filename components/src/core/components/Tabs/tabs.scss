@import '../../../styles/_colors.scss';


.tab-list {
    border: none;
    display: flex;
    border-bottom: 1px solid #dee2e6;
    padding: 0;
    .tab-items {
        cursor: pointer;
        width: fit-content;
        .tab-item{
          font-size: 0.875rem;
          color: $oxd-interface-gray-color;
          font-weight: 700;
          display: flex;
          padding: 0.5rem 1rem;
          margin-left: -10px;
          margin-right: 15px;
          margin-bottom: -5px;
          &.active {
            color: $oxd-interface-gray-darken-1-color;
          }
          &:hover, &:focus {
            background-color: $oxd-background-light-gray-color;
            border-radius: 2.5rem;
            color: $oxd-interface-gray-darken-1-color;
          }
          .tab-icon{
            font-size: 1.125rem;
            margin-right: 0.5rem;
          }
        }

        .selected-tab-indicator{
          margin-top: 8px;
          width: 30px;
          height: 4px;
          border-radius: 1.5rem;
          display: block;
          margin-bottom: -1px;
          &.active {
            background-color: $oxd-primary-one-color;
          }
        }

      }
    }

    ul {
      list-style-type: none;
    }



