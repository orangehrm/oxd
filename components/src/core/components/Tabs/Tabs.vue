<template>
  <ul class="tab-list">
    <li class="tab-items" v-for="(tab, i) in tabs" :key="i">
      <span
        tabindex="0"
        class="tab-item"
        @click="onClick($event, tab)"
        @focus="onFocus($event, tab)"
        @blur="onBlur($event, tab)"
        :class="{active: selectedTabId == tab.id}"
        @keyup.enter="onClick($event, tab)"
        :id="tab.id"
      >
        <oxd-icon class="tab-icon" :name="tab.icon" v-if="tab.icon" />{{
          $vt(tab.title)
        }}</span
      >
      <div
        class="selected-tab-indicator"
        :class="{active: selectedTabId == tab.id}"
      ></div>
    </li>
  </ul>
  <div class="tab-content">
    <slot></slot>
  </div>
</template>

<script lang="ts">
import {defineComponent, ref, provide} from 'vue';
import {Tab} from './types';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import translateMixin from '../../../mixins/translate';

export default defineComponent({
  name: 'oxd-tabs',
  components: {
    'oxd-icon': Icon,
  },
  mixins: [translateMixin],
  setup: function(props, {slots}) {
    const tabs = ref(
      slots.default
        ? slots
            .default()
            .filter(value => value.props?.tab)
            .map(value => {
              return {
                id: value.props?.tab.id,
                title: value.props?.tab.title,
                icon: value.props?.tab.icon,
              };
            })
        : [],
    );
    const selectedTabId = ref(((tabs as unknown) as any)?.value[0].id);
    provide('selectedTabId', selectedTabId);

    const setCurrentTabId = (tab: Tab): void => {
      selectedTabId.value = tab.id;
    };
    return {
      setCurrentTabId,
      selectedTabId,
      tabs,
    };
  },
  emits: ['click', 'focus', 'blur', 'change'],
  methods: {
    onClick($e: Event, tab: Tab) {
      this.$emit('click', $e, tab);
      this.$emit('change', $e, tab);
      this.setCurrentTabId(tab);
    },
    onFocus($e: Event, tab: Tab) {
      this.$emit('focus', $e, tab);
    },
    onBlur($e: Event, tab: Tab) {
      this.$emit('blur', $e, tab);
    },
  },
});
</script>

<style src="./tabs.scss" lang="scss" scoped></style>
