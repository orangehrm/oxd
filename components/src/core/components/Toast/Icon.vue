<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <div class="oxd-toast-icon-container">
    <div :class="iconWrapClasses">
      <oxd-icon :name="iconName" class="oxd-toast-icon" />
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {toastMixin} from './toast-mixin';
import {ToastType, ICON_MAP} from './types';
import Icon from '@ohrm/oxd/core/components/Icon/Icon.vue';

export default defineComponent({
  name: 'OxdToastIcon',

  components: {
    'oxd-icon': Icon,
  },

  mixins: [toastMixin],

  computed: {
    iconWrapClasses(): object {
      return {
        'oxd-toast-icon-wrap': true,
        [`oxd-toast-icon-wrap--${this.type}`]: true,
      };
    },
    iconName(): string {
      return ICON_MAP[this.type as ToastType];
    },
  },
});
</script>

<style src="./icon.scss" lang="scss" scoped></style>
