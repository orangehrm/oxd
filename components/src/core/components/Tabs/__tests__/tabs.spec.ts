import {mount} from '@vue/test-utils';
import Tabs from '@orangehrm/oxd/core/components/Tabs/Tabs.vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import {h} from 'vue';

describe('Tabs.vue', () => {
  it('Renders two tabs along with icons', async () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: [
          h(
            'Tab',
            {tab: {id: '1', title: 'Posts', icon: 'oxd-posts'}},
            ' This is the content of tab 1',
          ),
          h(
            'Tab',
            {tab: {id: '2', title: 'Posts', icon: 'oxd-profile'}},
            ' This is the content of tab2',
          ),
        ],
      },
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    const icon = wrapper.findComponent(Icon);
    expect(icon.find('.tab-icon')).toBeTruthy();
  });

  it('Renders two tabs without icons', async () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: [
          h(
            'Tab',
            {tab: {id: '1', title: 'Posts'}},
            ' This is the content of tab 1',
          ),
          h(
            'Tab',
            {tab: {id: '2', title: 'Posts'}},
            ' This is the content of tab2',
          ),
        ],
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.tab-icon').exists()).toBeFalsy();
  });

  it('In initial rendering first tab Id is set to the currentTabId', () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: [
          h(
            'Tab',
            {tab: {id: '1', title: 'Posts'}},
            ' This is the content of tab 1',
          ),
          h(
            'Tab',
            {tab: {id: '2', title: 'Posts'}},
            ' This is the content of tab2',
          ),
        ],
      },
    });
    expect(wrapper.vm.selectedTabId).toBe('1');
  });

  it('Click on the second tab', () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: [
          h(
            'Tab',
            {tab: {id: '1', title: 'Posts'}},
            ' This is the content of tab 1',
          ),
          h(
            'Tab',
            {tab: {id: 'tab2', title: 'Resume'}},
            ' This is the content of tab2',
          ),
        ],
      },
    });

    wrapper.find('#tab2').trigger('click');
    wrapper.vm.$nextTick();
    expect(wrapper.emitted().click[0]).toContainEqual({
      id: 'tab2',
      title: 'Resume',
    });

    expect(wrapper.emitted('change')).toBeTruthy();
    expect(wrapper.emitted('click')).toBeTruthy();
  });

  it('Hit Enter when focusing on the 2nd tab', () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: [
          h(
            'Tab',
            {tab: {id: 'tab1', title: 'Posts'}},
            ' This is the content of tab 1',
          ),
          h(
            'Tab',
            {tab: {id: 'tab2', title: 'Resume'}},
            ' This is the content of tab2',
          ),
        ],
      },
    });

    wrapper.find('#tab2').trigger('keyup.enter');
    wrapper.vm.$nextTick();
    expect(wrapper.emitted().change[0]).toContainEqual({
      id: 'tab2',
      title: 'Resume',
    });
    expect(wrapper.emitted('change')).toBeTruthy();
    expect(wrapper.emitted('click')).toBeTruthy();
  });

  it('Focusing on the third tab', () => {
    const wrapper = mount(Tabs, {
      props: {
        tabs: [
          {id: 'tab1', title: 'Details'},
          {id: 'tab2', title: 'Resume'},
          {id: 'tab3', title: 'Notes'},
        ],
      },
      slots: {
        default: [
          h(
            'Tab',
            {tab: {id: 'tab1', title: 'Details'}},
            ' This is the content of tab 1',
          ),
          h(
            'Tab',
            {tab: {id: 'tab2', title: 'Resume'}},
            ' This is the content of tab2',
          ),
          h(
            'Tab',
            {tab: {id: 'tab3', title: 'Notes'}},
            ' This is the content of tab2',
          ),
        ],
      },
    });

    wrapper.find('#tab3').trigger('focus');
    wrapper.vm.$nextTick();
    expect(wrapper.emitted().focus[0]).toContainEqual({
      id: 'tab3',
      title: 'Notes',
    });
    expect(wrapper.emitted('focus')).toBeTruthy();
  });

  it('Bluring on the first tab', () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: [
          h(
            'Tab',
            {tab: {id: 'tab1', title: 'Details'}},
            ' This is the content of tab 1',
          ),
          h(
            'Tab',
            {tab: {id: 'tab2', title: 'Resume'}},
            ' This is the content of tab2',
          ),
        ],
      },
    });

    wrapper.find('#tab1').trigger('blur');
    wrapper.vm.$nextTick();
    expect(wrapper.emitted().blur[0]).toContainEqual({
      id: 'tab1',
      title: 'Details',
    });
    expect(wrapper.emitted('blur')).toBeTruthy();
  });
});
