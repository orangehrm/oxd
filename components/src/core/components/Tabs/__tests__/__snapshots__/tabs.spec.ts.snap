// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.vue Renders two tabs along with icons 1`] = `
<ul class="tab-list">
  <li class="tab-items"><span tabindex="0" class="tab-item active" id="1"><i class="oxd-svg-icon oxd-svg-icon--medium tab-icon"><svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 516.07 512"><defs><style>.cls-1{fill:#64728c;}</style></defs><path class="cls-1" d="M388.07,512h-272a120.14,120.14,0,0,1-120-120V120a120.14,120.14,0,0,1,120-120h208c26.54,1.06,26.52,39,0,40h-208a80.09,80.09,0,0,0-80,80V392a80.09,80.09,0,0,0,80,80h272a80.09,80.09,0,0,0,80-80V207c1.06-26.54,39-26.52,40,0V392A120.13,120.13,0,0,1,388.07,512ZM266.77,266.34a20,20,0,0,1-19.63-23.61l10.47-56.19a2.38,2.38,0,0,1,.05-.26,143.77,143.77,0,0,1,39.48-73.78l90.73-91.26C415-7.07,463.8-7.08,491,21.2c28.26,27.12,28.26,75.92,0,103l-90.74,91.25A143.91,143.91,0,0,1,326.39,255l-55.78,11a20,20,0,0,1-3.84.37Zm27.5-69.67-5.06,27.14,26.86-5.27A104,104,0,0,0,369.38,190l90.74-91.26c12.82-12.3,12.83-34.32,0-46.62h0c-12.26-12.77-34.32-12.77-46.58,0l-90.74,91.26a104.12,104.12,0,0,0-28.57,53.3Z" transform="translate(3.93)"></path></svg></i>Posts</span>
    <div class="selected-tab-indicator active"></div>
  </li>
  <li class="tab-items"><span tabindex="0" class="tab-item" id="2"><i class="oxd-svg-icon oxd-svg-icon--medium tab-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 460.8 460.8" style="enable-background:new 0 0 460.8 460.8" xml:space="preserve"><path style="fill:currentColor" d="M230.432 0c-65.829 0-119.641 53.812-119.641 119.641s53.812 119.641 119.641 119.641 119.641-53.812 119.641-119.641S296.261 0 230.432 0zM435.755 334.89c-3.135-7.837-7.314-15.151-12.016-21.943-24.033-35.527-61.126-59.037-102.922-64.784-5.224-.522-10.971.522-15.151 3.657-21.943 16.196-48.065 24.555-75.233 24.555s-53.29-8.359-75.233-24.555c-4.18-3.135-9.927-4.702-15.151-3.657-41.796 5.747-79.412 29.257-102.922 64.784-4.702 6.792-8.882 14.629-12.016 21.943-1.567 3.135-1.045 6.792.522 9.927 4.18 7.314 9.404 14.629 14.106 20.898 7.314 9.927 15.151 18.808 24.033 27.167 7.314 7.314 15.673 14.106 24.033 20.898 41.273 30.825 90.906 47.02 142.106 47.02s100.833-16.196 142.106-47.02c8.359-6.269 16.718-13.584 24.033-20.898 8.359-8.359 16.718-17.241 24.033-27.167 5.224-6.792 9.927-13.584 14.106-20.898 2.611-3.135 3.133-6.793 1.566-9.927z"></path></svg></i>Posts</span>
    <div class="selected-tab-indicator"></div>
  </li>
</ul>
<div class="tab-content">
  <tab tab="[object Object]"> This is the content of tab 1</tab>
  <tab tab="[object Object]"> This is the content of tab2</tab>
</div>
`;
